_posts/vitalii_slobodianyk/2016-11-25-elasticsearch-shading.md

```diff
@@ -0,0 +1,350 @@
+---
+layout: post
+title:  "Shading Elasticsearch"
+breadcrumb: true
+author: vitalii_slobodianyk
+date: 2016-11-25
+categories: vitalii_slobodianyk
+tags:
+    - Elasticsearch
+    - Maven
+teaser:
+    info: A guide how to prepare an Elasticsearch distribution tailored to your needs.
+    image: vitalii_slobodianyk/elastic/elastic-logo.png
+header: 
+    version: small
+    title: Software Engineering Blog
+    image: header-logo-crop.png
+    icon: icon-blog
+---
+
+## Introduction
+
+'Shading' or package renaming(relocation) is a process of creating an [uber-jar](http://imagej.net/Uber-JAR) which contains its dependencies and package names of some of the dependendencies are renamed.
+
+In this blog post I will provide instructions how to create an Elasticsearch jar file with 'shaded' dependencies. This operation is often required when you use some libraries the Elasticsearch depends on, but version of the library differs. In such a case your project might not compile or run. A solution to the problem is to create a custom Elasticsearch distribution jar file which has the problematic dependencies 'shaded'(renamed).
+
+I will also guide how to create a 'shaded' jar file for testing with an in-memory Elasticsearch instance.
+
+## Maven project configuration
+
+We will use the Apache Maven to create the jar files, so let's prepare the standard maven directories layout.
+
+First, we create 2 modules which provide a shaded Elasticsearch artifact and shaded dependencies for Elasticsearch testing. A parent pom will be created to organize common configuration, versions and variables. The project layout will be the following:
+
+~~~
+.
+├── elasticsearch-shaded
+│   └── pom.xml
+├── elasticsearch-test-shaded
+│   └── pom.xml
+└── pom.xml
+
+2 directories, 3 files
+~~~
+
+#### Parent pom
+
+We start with a parent pom and put there all the common information we need as well as a `maven-shade-plugin` configuration. It easy to maintain plugin configuration if it is stored in one location.
+
+Define some common variables and settings we'll use:
+
+~~~xml
+<properties>
+  <!-- Versions -->
+  <elasticsearch.version>5.0.0</elasticsearch.version>
+  <shade-plugin.version>2.4.3</shade-plugin.version>
+  
+  <!-- Configuration -->
+  <elasticsearch.package>org.elasticsearch.shaded</elasticsearch.package>
+  <install.dir>${project.build.directory}/distribution</install.dir>
+  
+  <!-- Encoding -->
+  <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
+  <project.reporting.outputEncoding>UTF-8</project.reporting.outputEncoding>
+</properties>
+~~~
+
+Next, we will define shade plugin configuration. 
+
+We include only Elasticsearch packages and the relocated dependencies in shaded jar files. All the other Elasticsearch dependencies are promoted as transitive ones. Such a setup provides a better dependencies control, as it's possible to exclude some of the dependencies or override their version.
```

Comment:
"override their version." -> "override their version downstream in your project if the need arise"